baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/CardinalityEqualsZero.java

```diff
@@ -0,0 +1,135 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import java.util.Collection;
+import java.util.Optional;
+import org.immutables.value.Value.Immutable;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.SUGGESTION,
+        summary = "Use the isEmpty method instead of checking collection size")
+public final class CardinalityEqualsZero extends BugChecker implements BugChecker.BinaryTreeMatcher {
+    private static final Matcher<ExpressionTree> COLLECTION_SIZE_METHOD_MATCHER = MethodMatchers.instanceMethod()
+            .onDescendantOf(Collection.class.getName())
+            .named("size")
+            .withNoParameters();
+
+    private static final Matcher<ExpressionTree> INT_ZERO = Matchers.intLiteral(0);
+
+    @Override
+    public Description matchBinary(BinaryTree tree, VisitorState state) {
+        Optional<EqualsZeroExpression> maybeEqualsZeroExpression = getEqualsZeroExpression(tree, state);
+        if (maybeEqualsZeroExpression.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+
+        EqualsZeroExpression equalsZeroExpression = maybeEqualsZeroExpression.get();
+        ExpressionTree operand = equalsZeroExpression.operand();
+        ExpressionTree collectionInstance = ASTHelpers.getReceiver(operand);
+
+        if (collectionInstance == null || isExpressionThis(collectionInstance)) {
+            return Description.NO_MATCH;
+        }
+
+        if (COLLECTION_SIZE_METHOD_MATCHER.matches(operand, state)) {
+            return describeMatch(
+                    tree,
+                    SuggestedFix.replace(
+                            tree,
+                            (equalsZeroExpression.type() == ExpressionType.NEQ ? "!" : "")
+                                    + state.getSourceForNode(collectionInstance)
+                                    + ".isEmpty()"));
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private static boolean isExpressionThis(ExpressionTree tree) {
+        switch (tree.getKind()) {
+            case IDENTIFIER:
```

Comment:
I think the memberselect case is technically necessary here for "qualified this" usages e.g.
```java
class Outer {
  int foo();
  
  Inner getInner() {
    return new Inner();
  }
  class Inner {
    int foo() {
        // 'return foo()' would recurse infinitely
        // this allows us to reference components on an enclosing class from a non-static inner class.
        return Outer.this.foo();
    }
  }
}
```

Reply:
Ah. Have not seen this before. Updated.